### PR TITLE
Search API

### PR DESCRIPTION
**Pull request by nlake44**

### CODE REVIEW

File: AppController/lib/search.rb [hunk 1]
```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/ruby -w
+
+
+# First-party Ruby libraries
```
Comment (menivaitsi) [inline AppController/lib/search.rb]: Put '.' at the end. Same below.

File: AppController/lib/search.rb [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      Djinn.log_debug("Resuming from previous SOLR state if it exists.")
+    end
+
+    # First, start up SOLR.
+    start_solr()
+    # Start up the search server which handles API calls from applications.
+    start_search_server()
+  end
+
+  # Start up SOLR.
```
Comment (menivaitsi) [inline AppController/lib/search.rb]: Start -> Starts

File: AppController/lib/search.rb [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    Djinn.log_debug("Done stopping SOLR.")
+  end
+
+  # Stops SOLR and the search server on this node.
+  def self.stop()
+    self.stop_search_server()
+    self.stop_solr()
+  end
+
+
```
Comment (menivaitsi) [inline AppController/lib/search.rb]: Remove extra empty line. Same below.

File: AppServer/google/appengine/api/search/appscale_search_stub.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    self.__app_id = app_id
+ 
+  def _Dynamic_IndexDocument(self, request, response):
+    """ A local implementation of SearchService.IndexDocument RPC.
+
+    Index a new document or update an existing document.
+
+    Args:
+      request: A search_service_pb.IndexDocumentRequest.
+      response: An search_service_pb.IndexDocumentResponse.
```
Comment (menivaitsi) [inline AppServer/google/appengine/api/search/appscale_search_stub.py]: An -> A

Comment (menivaitsi) [inline AppServer/google/appengine/api/search/appscale_search_stub.py]: Same below.

File: AppServer/google/appengine/api/search/appscale_search_stub.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+  def Read(self):
+    """ Read search indexes from the index file.
+
+    This method is a no-op if index_file is set to None.
+    """
+    return
+
+  def _RemoteSend(self, request, response, method):
+    """ Sends a request remotely to the datstore server. """
```
Comment (menivaitsi) [inline AppServer/google/appengine/api/search/appscale_search_stub.py]: Please add an Args section.

File: InfrastructureManager/tests/test_ec2_agent.py [hunk 6]
```diff
@@ -71,8 +71,7 @@ def run_instances(self, prefix, blocking, success=True):
     # next, look at run_instances internally to make sure it actually is
     # updating its reservation info
     if not blocking:
-      time.sleep(2)
-
+      time.sleep(.1)
```
Comment (menivaitsi) [inline InfrastructureManager/tests/test_ec2_agent.py]: What is this change for?

File: InfrastructureManager/tests/test_infrastructure_manager_service.py [hunk 7]
```diff
@@ -22,7 +22,7 @@ def setUp(self):
      .should_receive('get_secret')
      .and_return('secret'))
     self.service, self.port = self.__start_service()
-    sleep(2)
+    #sleep(2)
```
Comment (menivaitsi) [inline InfrastructureManager/tests/test_infrastructure_manager_service.py]: Please remove dead code.

File: SearchService/query_parser.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+import urllib
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "../AppServer"))
+from google.appengine.api.search import query_parser
+from google.appengine.api.search import QueryParser
+
+# Encoded value for a comma ','.
+COMMA = "%2C"
+
+# Encoded value for a color ':'.
```
Comment (menivaitsi) [inline SearchService/query_parser.py]: color -> colon

File: SearchService/query_parser.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+      A str, the field list for the query.
+    """
+    field_string = ""
+    field_list = []
+    if self.__field_spec.name_size() > 0:
+      field_string += "&fl=id,"
+      for field_name in self.__field_spec.name_list():
+        field_list.append("{0}_{1}".format(self.__index.name, field_name))
+      field_string += SPACE.join(field_list) 
+      logging.debug("Field string: {0}".format(field_string))
```
Comment (menivaitsi) [inline SearchService/query_parser.py]: Please put a debug message similar to this one in __get_sort_list() above.

File: SearchService/query_parser.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+    """
+    for field in self.__index.schema.fields:
+      if field['name'].endswith(field_name) and \
+        field['name'].startswith("{0}_{1}_".format(self.__app_id,
+        self.__namespace)):
+        return field['name']
+    logging.error("Unable to find field name {0}".format(field_name))
+    return ""
+
+  def __dump_tree(self, node):
```
Comment (menivaitsi) [inline SearchService/query_parser.py]: Please add the Args section.

File: SearchService/query_parser.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+class Document(): 
+  """ Represents a document stored in SOLR. """ 
+ 
+  # Prefix code for index name. 
+  INDEX_NAME = "_gaeindex_name" 
+   
+  # Prefix code for index locale. 
+  INDEX_LOCALE = "_gaeindex_locale" 
+ 
+  def __init__(self, identifier, language, fields):
```
Comment (menivaitsi) [inline SearchService/query_parser.py]: Missing args section.

File: SearchService/search_api.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
+""" Class for handling searialized Search requests. """
```
Comment (menivaitsi) [inline SearchService/search_api.py]: searialized -> serialized

File: SearchService/search_api.py [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+    return apiresponse.Encode()
+
+  def index_document(self, data):
+    """ Index a new document or update an existing document.
+ 
+    Args:
+      data: A str. Serialized protocol buffer.
+    Returns:
+      A tuple of a encoded response, error code, and error detail.
```
Comment (menivaitsi) [inline SearchService/search_api.py]: "a encoded" -> "an encoded"

File: SearchService/search_api.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+        new_status.set_code(
+          search_service_pb.SearchServiceError.INTERNAL_ERROR)
+
+    return response.Encode(), 0, ""
+
+  def delete_document(self, data):
+    """ Deletes a document.
+ 
+    Args:
+      data: A str. Serialize protocol buffer.
```
Comment (menivaitsi) [inline SearchService/search_api.py]: Serialize -> Serialized

File: SearchService/search_api.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    Returns:
+      A tuple of a encoded response, error code, and error detail.
```
Comment (menivaitsi) [inline SearchService/search_api.py]: "a encoded" -> "an encoded"

Comment (menivaitsi) [inline SearchService/search_api.py]: Same below

File: SearchService/search_api.py (hunk 14 above)
Comment (menivaitsi) [inline SearchService/search_api.py]: Same below

File: SearchService/solr_interface.py [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+      A str, the internal name of the index.
+    """
+    return app_id + "_" + namespace + "_" + name  
+
+  def delete_doc(self, doc_id):
+    """ Deletes a document by doc ID.
+
+    Args:
+      doc_id: A list of document IDs.
+    """
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Put a Raises section with the appropriate exception.

File: SearchService/solr_interface.py [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+    solr_request = {"delete": {"id": doc_id}}
+    solr_url = "http://{0}:{1}/solr/update?".format(self._search_location,
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Here's a suggestion:
Make this portion: "http://{0}:{1}/solr/" a constant and append the paths ('/update?...', '/schema/fields, etc.') when needed.

File: SearchService/solr_interface.py [hunk 18]
```diff
@@ -0,0 +1,491 @@
...
+      self.SOLR_SERVER_PORT)
+    logging.info("SOLR URL: {0}".format(solr_url))
+    json_request = simplejson.dumps(solr_request)
+    logging.info("SOLR JSON: {0}".format(json_request))
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Should this be a debug message?

File: SearchService/solr_interface.py [hunk 19]
```diff
@@ -0,0 +1,491 @@
...
+    try:
+      req = urllib2.Request(solr_url, data=json_request)
+      req.add_header('Content-Type', 'application/json')
+      conn = urllib2.urlopen(req) 
+      if conn.getcode() != 200:
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Create constants for HTTP codes please.

File: SearchService/solr_interface.py [hunk 20]
```diff
@@ -0,0 +1,491 @@
...
+        raise search_exceptions.InternalError("Malformed response from SOLR.")
+      response = simplejson.load(conn)
+      status = response['responseHeader']['status'] 
+      logging.info("Response: {0}".format(response))
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Make this logging.debug?

File: SearchService/solr_interface.py [hunk 21]
```diff
@@ -0,0 +1,491 @@
...
+        filtered_fields.append(field)
+    schema = Schema(filtered_fields, response['responseHeader'])
+    return Index(index_name, schema)
+
+  def update_schema(self, updates):
+    """ Updates the schema of a document.
+
+    Args:
+      updates: A list of updates to apply.
+    """
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Missing Raises section.

File: SearchService/solr_interface.py [hunk 22]
```diff
@@ -0,0 +1,491 @@
...
+      doc: The document to update.
+      index_spec: An index specification.
+    """
+    solr_doc = self.to_solr_doc(doc)
+
+    index = self.get_index(app_id, index_spec.namespace(), index_spec.name())
+    updates = self.compute_updates(index.name, index.schema.fields,
+      solr_doc.fields)
+    if len(updates) > 0:
+      self.update_schema(updates)
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: What happens if update_schema fails? Put a try block here.

File: SearchService/solr_interface.py [hunk 23]
```diff
@@ -0,0 +1,491 @@
...
+    self.commit_update(hash_map)
+
+  def to_solr_doc(self, doc):
+    """ Converts to an internal SOLR document. 
+
+    Args:
+      doc: A document_pb.Document type.
+    Returns:
+      A converted Document type.
+    """
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Missing Raises section.

File: SearchService/solr_interface.py [hunk 24]
```diff
@@ -0,0 +1,491 @@
...
+      search_params.offset())
+    solr_query = parser.get_solr_query_string(query)
+    logging.debug("Solr query: {0}".format(solr_query))
+    solr_results = self.__execute_query(solr_query)
+    logging.debug("Solr results: {0}".format(solr_results))
+    self.__convert_to_gae_results(result, solr_results, index)
+    logging.debug("GAE results: {0}".format(result))
+
+  def __execute_query(self, solr_query):
+    """ Executes query string on SOLR. """
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Missing Args, Returns and Raises sections.

File: SearchService/solr_interface.py [hunk 25]
```diff
@@ -0,0 +1,491 @@
...
+        "SOLR response status of {0}".format(status))
+    return response
+
+  def __convert_to_gae_results(self, result, solr_results, index):
+    """ Converts SOLR results in to GAE compatible documents. 
+
+    Args:
+      result: A search_service_pb.SearchResponse.
+      solr_results: A dictionary returned from SOLR on a search query.
+      namespace: A str, the namespace.
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: There is not such argument here.

File: SearchService/solr_interface.py [hunk 26]
```diff
@@ -0,0 +1,491 @@
...
+      if field_type == "":
+        logging.warning("Unable to find type for {0}").format(
+          "{0}_{1}".format(index.name, field_name))
+      self.__add_field_value(new_value, doc[key], field_type)
+
+  def __add_field_value(self, new_value, value, ftype):
+    """ Adds a value to a result field.
+
+    Args:
+      new_value:
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Missing arg description.

File: SearchService/solr_interface.py [hunk 27]
```diff
@@ -0,0 +1,491 @@
...
+    else:
+      logging.warning("Default field found! {0}".format(ftype))
+      new_value.set_string_value(value)
+      new_value.set_type(FieldValue.TEXT)
+    logging.debug("New value: {0}".format(new_value))
+
+class Schema():
+  """ Represents a schema in SOLR. """
+  def __init__(self, fields , response_header):
+    """ Constructor for SOLR schema. """
```
Comment (menivaitsi) [inline SearchService/solr_interface.py]: Missing Args section.

Comment (menivaitsi) [inline SearchService/solr_interface.py]: Same below.

File: SearchService/test/unit/test_solr_interface.py [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+    self.assertRaises(search_exceptions.InternalError, solr.commit_update, {})
+
+    dictionary = {'responseHeader':{'status': 0}}
+    simplejson.should_receive("load").and_return(dictionary).once()
+    solr.commit_update({})
+
+  def test_update_document(self):
+    appscale_info = flexmock()
+    appscale_info.should_receive("get_search_location").and_return("somelocation")
+    solr = solr_interface.Solr()
```
Comment (menivaitsi) [inline SearchService/test/unit/test_solr_interface.py]: Remove dup line.

**Comment by menivaitsi** [comment]
Good job!

File: InfrastructureManager/tests/test_ec2_agent.py (hunk 6 above)
Comment (nlake44) [inline InfrastructureManager/tests/test_ec2_agent.py]: Makes the unit tests faster.

**Comment by menivaitsi** [comment]
:+1:

**Pull request merged**